### PR TITLE
#Network New cmdlets for Trusted Root Certificate and Autoscale Configuration for Application Gateway 

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Description
This PR adds Trusted Root Certificate and Autoscale Configuration.

https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/156
Needs to be made public before Ignite.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@akshaysngupta To fix the current Jenkins error, please follow instructions here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/7166/artifact/src/Package/ValidateHelpIssues.csv/*view*/

**Comment by cormacpayne** [comment]
@akshaysngupta please make sure that the version of the Network SDK being used is updated across all projects using it (Compute, Dns, RecoveryServices, Sql)

**Comment by maddieclayton** [comment]
@akshaysngupta please link your cmdlet design review from here: https://github.com/Azure/azure-powershell-cmdlet-review-pr

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/AddAzureApplicationGatewayTrustedRootCertificateCommand.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Add, "AzureRmApplicationGatewayTrustedRootCertificate", SupportsShouldProcess = true),
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/AddAzureApplicationGatewayTrustedRootCertificateCommand.cs]: Kiindly use `ResourceManager.Common.AzureRMConstants.AzureRMPrefix` for the AzureRm part of the prefix to the noun part. Example [here in the current PR itself](https://github.com/Azure/azure-powershell/pull/7121/files#diff-f36fcbd8e47bf79727d50081d50419e6R21). This applies to all instances where the name of the cmdlet is defined in the PR.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/GetAzureApplicationGatewayTrustedRootCertificateCommand.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmApplicationGatewayTrustedRootCertificate"),
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/GetAzureApplicationGatewayTrustedRootCertificateCommand.cs]: Kiindly use `ResourceManager.Common.AzureRMConstants.AzureRMPrefix` for the AzureRm part of the prefix to the noun part. Example [here in the current PR itself](https://github.com/Azure/azure-powershell/pull/7121/files#diff-f36fcbd8e47bf79727d50081d50419e6R21). This applies to all instances where the name of the cmdlet is defined in the PR.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/NewAzureApplicationGatewayTrustedRootCertificateCommand.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, "AzureRmApplicationGatewayTrustedRootCertificate", SupportsShouldProcess = true),
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/NewAzureApplicationGatewayTrustedRootCertificateCommand.cs]: Kiindly use `ResourceManager.Common.AzureRMConstants.AzureRMPrefix` for the AzureRm part of the prefix to the noun part. Example [here in the current PR itself](https://github.com/Azure/azure-powershell/pull/7121/files#diff-f36fcbd8e47bf79727d50081d50419e6R21). This applies to all instances where the name of the cmdlet is defined in the PR.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/RemoveAzureApplicationGatewayTrustedRootCertificateCommand.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Remove, "AzureRmApplicationGatewayTrustedRootCertificate", SupportsShouldProcess = true),
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/RemoveAzureApplicationGatewayTrustedRootCertificateCommand.cs]: Kiindly use `ResourceManager.Common.AzureRMConstants.AzureRMPrefix` for the AzureRm part of the prefix to the noun part. Example [here in the current PR itself](https://github.com/Azure/azure-powershell/pull/7121/files#diff-f36fcbd8e47bf79727d50081d50419e6R21). This applies to all instances where the name of the cmdlet is defined in the PR.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/SetAzureApplicationGatewayTrustedRootCertificateCommand.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+using Microsoft.Azure.Commands.Network.Models;
+using System;
+using System.Linq;
+using System.Management.Automation;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.Set, "AzureRmApplicationGatewayTrustedRootCertificate", SupportsShouldProcess = true),
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/TrustedRootCertificate/SetAzureApplicationGatewayTrustedRootCertificateCommand.cs]: Kiindly use `ResourceManager.Common.AzureRMConstants.AzureRMPrefix` for the AzureRm part of the prefix to the noun part. Example [here in the current PR itself](https://github.com/Azure/azure-powershell/pull/7121/files#diff-f36fcbd8e47bf79727d50081d50419e6R21). This applies to all instances where the name of the cmdlet is defined in the PR.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/BackendHttpSettings/AzureApplicationGatewayBackendHttpSettingsBase.cs [hunk 6]
```diff
@@ -78,6 +78,12 @@ public class AzureApplicationGatewayBackendHttpSettingsBase : NetworkBaseCmdlet
         [ValidateNotNullOrEmpty]
         public List<PSApplicationGatewayAuthenticationCertificate> AuthenticationCertificates { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "Application gateway Trusted Root Certificates")]
+        [ValidateNotNullOrEmpty]
+        public List<PSApplicationGatewayTrustedRootCertificate> TrustedRootCertificates { get; set; }
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/BackendHttpSettings/AzureApplicationGatewayBackendHttpSettingsBase.cs]: why is this a plural? the param name should be singular i.e `TrustedRootCertificate`. [Documentation](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/azure-powershell-design-guidelines.md#noun-singularity).

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 7]
```diff
@@ -83,6 +83,12 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
              HelpMessage = "The list of authentication certificates")]
         public List<PSApplicationGatewayAuthenticationCertificate> AuthenticationCertificates { get; set; }
 
+        [Parameter(
+             Mandatory = false,
+             ValueFromPipelineByPropertyName = true,
+             HelpMessage = "The list of trusted root certificates")]
+        public List<PSApplicationGatewayTrustedRootCertificate> TrustedRootCertificates { get; set; }
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs]: why is this a plural? the param name should be singular i.e `TrustedRootCertificate`. [Documentation](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/azure-powershell-design-guidelines.md#noun-singularity).

**Comment by EvgenyAgafonchikov** [comment]
@cormacpayne, @akshaysngupta, I believe merging [this](https://github.com/Azure/azure-powershell/pull/7202) and rebasing current PR should cover updating NRP SDK version in required projects and fix recordings, so we get "clean" list of errors (currently there are some unrelated to this PR as missing recordings)

**Comment by EvgenyAgafonchikov** [comment]
@akshaysngupta, AFAIK [these](https://github.com/Azure/azure-powershell/pull/7121/files#diff-b60b2559dbc8dde377acf5fe189d4668) changes should be also ported [there](https://github.com/Azure/azure-powershell/blob/network-september-release/src/ResourceManager/Network/Commands.Network/Az.Network.psd1)
And the next step is to regenerate the aliases using `tools/CreateAliasMapping` and copy over the output (found in tools/AliasMappings.json), to the mapping [here](https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-powershell%2Fblob%2Fpreview%2Fsrc%2FResourceManager%2FProfile%2FCommands.Profile%2FAzureRmAlias%2FMappings.cs%23L36&data=02%7C01%7C%7Cb506b546334b41f65e6008d60ea2bfcf%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636712489492421714&sdata=A61nBKhYJcKr4fSkgdi9%2BOcdb0r4AqLJupk6Ugsd98U%3D&reserved=0).
@maddieclayton, please confirm this flow now is like described.

**Comment by maddieclayton** [comment]
@EvgenyAgafonchikov @akshaysngupta Yes, that flow is correct.  You can find the doc for this here: https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/create-alias-mappings.md

**Comment by akshaysngupta** [comment]
Thanks for feedback all. I have updated the PR with the latest changes.

File: src/ResourceManager/Network/Commands.Network/ChangeLog.md [hunk 8]
```diff
@@ -22,6 +22,25 @@
 
 ## Version 6.7.0
 * Updated cmdlet New-AzureRmVirtualNetworkGatewayConnection with support for switch ExpressRouteGatewayBypass
+* Added support for Trusted Root certificate and Autoscale configuration in Application Gateway
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ChangeLog.md]: you need to move these under `Current Release`

File: src/ResourceManager/Network/Commands.Network/ChangeLog.md [hunk 9]
```diff
@@ -22,6 +22,25 @@
...
+        - Add-AzureRmApplicationGatewayTrustedRootCertificate
+        - Get-AzureRmApplicationGatewayTrustedRootCertificate
+        - New-AzureRmApplicationGatewayTrustedRootCertificate
+        - Remove-AzureRmApplicationGatewayTrustedRootCertificate
+        - Set-AzureRmApplicationGatewayTrustedRootCertificate
+        - Get-AzureRmApplicationGatewayAutoscaleConfiguration
+        - New-AzureRmApplicationGatewayAutoscaleConfiguration
+        - Remove-AzureRmApplicationGatewayAutoscaleConfiguration
+        - Set-AzureRmApplicationGatewayAutoscaleConfiguration
+    - Cmdlets updated with optonal parameter -TrustedRootCertificates
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ChangeLog.md]: `TrustedRootCertificate`  instead of `TrustedRootCertificates`

**Comment by EvgenyAgafonchikov** [comment]
CI passed. @maddieclayton, @praries880, could you please take a look?

File: src/ResourceManager/Network/Commands.Network/ChangeLog.md [hunk 10]
```diff
@@ -41,6 +41,28 @@
...
+    - Cmdlets updated with optonal parameter -TrustedRootCertificate
+        - New-AzureRmApplicationGateway
+        - Set-AzureRmApplicationGateway
+        - New-AzureRmApplicationGatewayBackendHttpSetting
+        - Set-AzureRmApplicationGatewayBackendHttpSetting
+    - Cmdlets updated with optonal parameter -AutoscaleConfiguration
+        - New-AzureRmApplicationGateway
+        - Set-AzureRmApplicationGateway
+
+## Version 6.7.0
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ChangeLog.md]: remove the duplicated 6.7.0 header

**Comment by akshaysngupta** [comment]
Corrected.

**Pull request merged**